testing_utilities.go

```diff
@@ -0,0 +1,359 @@
+
+		if ionHashReader.isInStruct() {
+			compareFieldNames(t, reader1, reader2)
+		}
+
+		compareAnnotations(t, reader1, reader2)
+
+		compareAnnotationSymbols(t, reader1, reader2)
+
+		compareHasAnnotations(t, reader1, reader2)
+
+		require.Equal(t, reader1.IsNull(), reader2.IsNull(), "Expected readers to have matching IsNull() values")
+
+		switch reader1.Type() {
+		case ion.NullType:
+			assert.True(t, reader1.IsNull(), "Expected reader1.IsNull() to return true")
+			assert.True(t, reader2.IsNull(), "Expected reader2.IsNull() to return true")
+		case ion.BoolType, ion.IntType, ion.FloatType, ion.DecimalType, ion.TimestampType,
+			ion.StringType, ion.SymbolType, ion.BlobType, ion.ClobType:
+
+			compareScalars(t, reader1.Type(), reader1.IsNull(), reader1, reader2)
+		case ion.StructType, ion.ListType, ion.SexpType:
+			assert.NoError(t, reader1.StepIn(), "Something went wrong executing reader1.StepIn()")
+
+			assert.NoError(t, reader2.StepIn(), "Something went wrong executing reader2.StepIn()")
+
+			compareReaders(t, reader1, reader2)
+
+			assert.NoError(t, reader1.StepOut(), "Something went wrong executing reader1.StepOut()")
+
+			assert.NoError(t, reader2.StepOut(), "Something went wrong executing reader2.StepOut()")
+		default:
+			t.Error(&InvalidIonTypeError{reader1.Type()})
+		}
+	}
+
+	assert.False(t, hasNext(t, reader1, reader2), "Expected hasNext() to return false")
+}
+
+// hasNext() checks that the readers have a Next value
+func hasNext(t *testing.T, reader1 ion.Reader, reader2 ion.Reader) bool {
+	next1 := reader1.Next()
+	next2 := reader2.Next()
+
+	assert.Equal(t, next1, next2, "next results don't match")
+
+	if !next1 {
+		assert.NoError(t, reader1.Err(), "Something went wrong executing reader1.next()")
+	}
+
+	if !next2 {
+		assert.NoError(t, reader2.Err(), "Something went wrong executing reader2.next()")
+	}
+
+	return next1 && next2
+}
+
+func compareFieldNames(t *testing.T, reader1 ion.Reader, reader2 ion.Reader) {
+	// TODO: Add SymbolToken logic here once SymbolTokens are available
+}
+
+func compareNonNullStrings(t *testing.T, str1, str2 string) {
+	assert.NotNil(t, str1, "Expected str1 to be not null")
+	assert.NotNil(t, str2, "Expected str2 to be not null")
+	assert.Equal(t, str1, str2, "Expected strings to match")
+}
+
+func compareAnnotations(t *testing.T, reader1 ion.Reader, reader2 ion.Reader) {
+	// TODO: Add SymbolToken logic here once SymbolTokens are available
+
+	assert.Equal(t, reader1.Annotations(), reader2.Annotations(), "Expected symbol sequences to match")
+}
+
+func compareAnnotationSymbols(t *testing.T, reader1, reader2 ion.Reader) {
+	// TODO: Add SymbolToken logic here once SymbolTokens are available
+}
+
+func compareHasAnnotations(t *testing.T, reader1, reader2 ion.Reader) {
+	// TODO: Add SymbolToken logic here once SymbolTokens are available
+}
+
+func compareScalars(t *testing.T, ionType ion.Type, isNull bool, reader1 ion.Reader, reader2 ion.Reader) {
```

Comment:
I recall seeing something similar to what we have here, the `isNull` argument. Why are we excluding comparisons of Ion null? It is a scalar and would make this utility function "complete".

Reply:
This function now compares the readers' IsNull() values.